File: enhanced_demo.js
#!/usr/bin/env node

/**
 * Enhanced Timestamp Compression Demo
 *
 * This script demonstrates the additional compression methods from enhanced_compression.js
 * and compares them with the existing methods.
 */

// Import the enhanced compression methods
const enhancedCompression = require('./enhanced_compression.js');

// Define the base62 function directly to avoid import issues
function compressMillisBase62(millis) {
  const chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
  const base = chars.length; // 62

  // Convert to Base62
  let result = "";
  let number = millis;

  do {
    result = chars[number % base] + result;
    number = Math.floor(number / base);
  } while (number > 0);

  return result;
}

/**
 * Calculate size information about a representation
 * @param {number} original - Original milliseconds value
 * @param {string} representation - The encoded representation
 * @returns {object} Size and efficiency information
 */
function getSizeInfo(original, representation) {
  const originalBytes = original.toString().length;
  const representationBytes = representation.toString().length;
  const saving = originalBytes - representationBytes;
  const savingPercent = originalBytes > 0 ? (saving / originalBytes) * 100 : 0;

  let efficiency;
  if (saving >= 0) {
    efficiency = `Saved ${saving} characters (${savingPercent.toFixed(2)}%)`;
  } else {
    efficiency = `Increased by ${Math.abs(saving)} characters (${Math.abs(savingPercent).toFixed(2)}%)`;
  }

  return {
    originalSize: originalBytes,
    representationSize: representationBytes,
    efficiency
  };
}

/**
 * Process a timestamp with enhanced compression methods
 * @param {number} millis - milliseconds since epoch
 */
function processTimestampEnhanced(millis) {
  console.log(`Original timestamp: ${millis} milliseconds since epoch`);
  console.log("-".repeat(50));

  // Existing Base62 method for comparison
  const base62Repr = compressMillisBase62(millis);
  const base62Info = getSizeInfo(millis, base62Repr);
  console.log(`Base62 representation (existing): ${base62Repr}`);
  console.log(`Size: ${base62Info.representationSize} chars, ${base62Info.efficiency}`);
  console.log("-".repeat(50));

  // Enhanced Method 1: Base94 compression
  const base94Repr = enhancedCompression.compressMillisBase94(millis);
  const base94Info = getSizeInfo(millis, base94Repr);
  console.log(`Base94 representation: ${base94Repr}`);
  console.log(`Size: ${base94Info.representationSize} chars, ${base94Info.efficiency}`);
  console.log("-".repeat(50));

  // Enhanced Method 2: Variable-length compression
  const variableRepr = enhancedCompression.compressMillisVariable(millis);
  const variableInfo = getSizeInfo(millis, variableRepr);
  console.log(`Variable-length representation: ${variableRepr}`);
  console.log(`Size: ${variableInfo.representationSize} chars, ${variableInfo.efficiency}`);
  console.log("-".repeat(50));

  // Enhanced Method 3: Bit-packed compression
  const bitPackedRepr = enhancedCompression.compressMillisBitPacked(millis);
  const bitPackedInfo = getSizeInfo(millis, bitPackedRepr);
  console.log(`Bit-packed representation: ${bitPackedRepr}`);
  console.log(`Size: ${bitPackedInfo.representationSize} chars, ${bitPackedInfo.efficiency}`);
  console.log("-".repeat(50));

  // Enhanced Method 4: Delta encoding (using a series for demonstration)
  const series = [millis, millis + 5000, millis + 12000];
  const seriesRepr = enhancedCompression.compressTimestampSeries(series);
  console.log("Timestamp Series:");
  console.log(`Original series: [${series.join(", ")}]`);
  console.log(`Delta encoded: ${seriesRepr}`);
  console.log(`Size: ${seriesRepr.length} chars for 3 timestamps (avg: ${(seriesRepr.length / 3).toFixed(2)} per timestamp)`);
  console.log("-".repeat(50));

  // Summary
  console.log("Summary of enhanced representations:");
  console.log(`1. Original:          ${millis} (${millis.toString().length} chars)`);
  console.log(`2. Base62 (existing): ${base62Repr} (${base62Info.representationSize} chars)`);
  console.log(`3. Base94:            ${base94Repr} (${base94Info.representationSize} chars)`);
  console.log(`4. Variable-length:   ${variableRepr} (${variableInfo.representationSize} chars)`);
  console.log(`5. Bit-packed:        ${bitPackedRepr} (${bitPackedInfo.representationSize} chars)`);
}

/**
 * Main function to handle CLI arguments
 */
function main() {
  // Get milliseconds from command line arguments or use current time
  let millis;

  if (process.argv.length > 2) {
    millis = parseInt(process.argv[2], 10);
    if (isNaN(millis)) {
      console.error("Error: Please provide a valid integer for milliseconds.");
      process.exit(1);
    }
  } else {
    // Use current time if no argument provided
    millis = Date.now();
    console.log("No timestamp provided. Using current time.");
  }

  processTimestampEnhanced(millis);
}

// Run the main function
main();

File: enhanced_demo.py
#!/usr/bin/env python3

"""
Enhanced Timestamp Compression Demo

This script demonstrates the additional compression methods from enhanced_compression.py
and compares them with the existing methods.
"""

import sys
import time
from time_converter import compress_millis_base62
from enhanced_compression import (
    compress_millis_base94,
    compress_millis_variable,
    compress_millis_bit_packed,
    compress_timestamp_series
)

def get_size_info(original, representation):
    """Calculate size information about a representation."""
    original_bytes = len(str(original))
    representation_bytes = len(representation)
    saving = original_bytes - representation_bytes
    saving_percent = (saving / original_bytes) * 100 if original_bytes > 0 else 0

    if saving >= 0:
        efficiency = f"Saved {saving} characters ({saving_percent:.2f}%)"
    else:
        efficiency = f"Increased by {abs(saving)} characters ({abs(saving_percent):.2f}%)"

    return {
        "original_size": original_bytes,
        "representation_size": representation_bytes,
        "efficiency": efficiency
    }

def process_timestamp_enhanced(millis):
    """Process a timestamp with enhanced compression methods."""
    print(f"Original timestamp: {millis} milliseconds since epoch")
    print("-" * 50)

    # Existing Base62 method for comparison
    base62_repr = compress_millis_base62(millis)
    base62_info = get_size_info(millis, base62_repr)
    print(f"Base62 representation (existing): {base62_repr}")
    print(f"Size: {base62_info['representation_size']} chars, {base62_info['efficiency']}")
    print("-" * 50)

    # Enhanced Method 1: Base94 compression
    base94_repr = compress_millis_base94(millis)
    base94_info = get_size_info(millis, base94_repr)
    print(f"Base94 representation: {base94_repr}")
    print(f"Size: {base94_info['representation_size']} chars, {base94_info['efficiency']}")
    print("-" * 50)

    # Enhanced Method 2: Variable-length compression
    variable_repr = compress_millis_variable(millis)
    variable_info = get_size_info(millis, variable_repr)
    print(f"Variable-length representation: {variable_repr}")
    print(f"Size: {variable_info['representation_size']} chars, {variable_info['efficiency']}")
    print("-" * 50)

    # Enhanced Method 3: Bit-packed compression
    bit_packed_repr = compress_millis_bit_packed(millis)
    bit_packed_info = get_size_info(millis, bit_packed_repr)
    print(f"Bit-packed representation: {bit_packed_repr}")
    print(f"Size: {bit_packed_info['representation_size']} chars, {bit_packed_info['efficiency']}")
    print("-" * 50)

    # Enhanced Method 4: Delta encoding (using a series for demonstration)
    series = [millis, millis + 5000, millis + 12000]
    series_repr = compress_timestamp_series(series)
    print("Timestamp Series:")
    print(f"Original series: {series}")
    print(f"Delta encoded: {series_repr}")
    print(f"Size: {len(series_repr)} chars for 3 timestamps (avg: {len(series_repr)/3:.2f} per timestamp)")
    print("-" * 50)

    # Summary
    print("Summary of enhanced representations:")
    print(f"1. Original:          {millis} ({len(str(millis))} chars)")
    print(f"2. Base62 (existing): {base62_repr} ({base62_info['representation_size']} chars)")
    print(f"3. Base94:            {base94_repr} ({base94_info['representation_size']} chars)")
    print(f"4. Variable-length:   {variable_repr} ({variable_info['representation_size']} chars)")
    print(f"5. Bit-packed:        {bit_packed_repr} ({bit_packed_info['representation_size']} chars)")

def main():
    if len(sys.argv) > 1:
        try:
            millis = int(sys.argv[1])
            process_timestamp_enhanced(millis)
        except ValueError:
            print("Error: Please provide a valid integer for milliseconds.")
            sys.exit(1)
    else:
        # Use current time if no argument provided
        current_millis = int(time.time() * 1000)
        print("No timestamp provided. Using current time.")
        process_timestamp_enhanced(current_millis)

if __name__ == "__main__":
    main()
